euphoria-core/src/main/java/cz/seznam/euphoria/core/executor/io/GenericSpillTools.java

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.seznam.euphoria.core.executor.io;
+
+import cz.seznam.euphoria.core.annotation.audience.Audience;
+import cz.seznam.euphoria.core.client.io.ExternalIterable;
+import cz.seznam.euphoria.core.client.io.SpillTools;
+import cz.seznam.euphoria.core.executor.Constants;
+import cz.seznam.euphoria.core.util.Settings;
+import cz.seznam.euphoria.shaded.guava.com.google.common.annotations.VisibleForTesting;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * An implementation of {@code SpillTools} to be used by executors.
+ */
+@Audience(Audience.Type.EXECUTOR)
+public class GenericSpillTools implements SpillTools {
+
+  /**
+   * Number of records to keep in list before spilling.
+   */
+  private final int spillRecords;
+
+  /**
+   * Factory for creating files for spilling.
+   */
+  private final SpillFileFactory spillFactory;
+
+  /**
+   * Factory for serializer used when spilling data to disk.
+   */
+  private final SerializerFactory serializer;
+
+  /**
+   * @param serializer path to temporary directory to use for spilling
+   * @param spillFactory factory for creating files for spilling
+   * @param settings settings to read configuration from
+   */
+  public GenericSpillTools(
+      SerializerFactory serializer,
+      SpillFileFactory spillFactory,
+      Settings settings) {
+
+    this(serializer, spillFactory, settings.getInt(
+        Constants.SPILL_BUFFER_ITEMS, Constants.SPILL_BUFFER_ITEMS_DEFAULT));
+  }
+
+  /**
+   * @param serializer path to temporary directory to use for spilling
+   * @param settings settings to read configuration from
+   */
+  public GenericSpillTools(
+      SerializerFactory serializer,
+      Settings settings) {
+
+    this(serializer, spillFactory(settings), settings.getInt(
+        Constants.SPILL_BUFFER_ITEMS, Constants.SPILL_BUFFER_ITEMS_DEFAULT));
+  }
+
+  @VisibleForTesting
+  GenericSpillTools(
+      SerializerFactory serializer,
+      SpillFileFactory spillFactory,
+      int spillRecords) {
+
+    this.serializer = serializer;
+    this.spillFactory = spillFactory;
+    this.spillRecords = spillRecords;
+  }
+
+  private static SpillFileFactory spillFactory(Settings settings) {
+    final File tmpDir = new File(settings.getString(
+        Constants.LOCAL_TMP_DIR,
+        Constants.LOCAL_TMP_DIR_DEFAULT));
+
+    if (tmpDir.exists()) {
+      if (!tmpDir.isDirectory()) {
+        throw new IllegalArgumentException(
+            "Path " + tmpDir
+            + " exists and is not directory! Tune your " + Constants.LOCAL_TMP_DIR
+            + " settings");
+      }
+    } else {
+      tmpDir.mkdirs();
```

Comment:
I'd say no, because that would mean that the `tmpDir.exists()` would return `true`. There could be of course a race condition, that the directory would be created after call to `tmpDir.exists()` and before `tmpDir.mkdirs()`, but either way, return false would just mean that the directories have not been created, because it already exists. There is probably no problem with that.